README.md
<div align="center">
  <a href="https://piotrmurach.github.io/tty" target="_blank"><img width="130" src="https://github.com/piotrmurach/tty/raw/master/images/tty.png" alt="tty logo" /></a>
</div>

# TTY::Pager [![Gitter](https://badges.gitter.im/Join%20Chat.svg)][gitter]

[![Gem Version](https://badge.fury.io/rb/tty-pager.svg)][gem]
[![Build Status](https://secure.travis-ci.org/piotrmurach/tty-pager.svg?branch=master)][travis]
[![Build status](https://ci.appveyor.com/api/projects/status/3auc1vi3mk5puqai?svg=true)][appveyor]
[![Maintainability](https://api.codeclimate.com/v1/badges/636da0d02231b7f3e50f/maintainability)][codeclimate]
[![Coverage Status](https://coveralls.io/repos/github/piotrmurach/tty-pager/badge.svg)][coverage]
[![Inline docs](http://inch-ci.org/github/piotrmurach/tty-pager.svg?branch=master)][inchpages]

[gitter]: https://gitter.im/piotrmurach/tty
[gem]: http://badge.fury.io/rb/tty-pager
[travis]: http://travis-ci.org/piotrmurach/tty-pager
[appveyor]: https://ci.appveyor.com/project/piotrmurach/tty-pager
[codeclimate]: https://codeclimate.com/github/piotrmurach/tty-pager/maintainability
[coverage]: https://coveralls.io/github/piotrmurach/tty-pager
[inchpages]: http://inch-ci.org/github/piotrmurach/tty-pager

> A cross-platform terminal pager that works on all major Ruby interpreters.

**TTY::Pager** provides independent terminal pager component for [TTY](https://github.com/piotrmurach/tty) toolkit.

## Installation

Add this line to your application's Gemfile:

    gem 'tty-pager'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install tty-pager


## Overview

The **TTY::Pager** will automatically choose the best available pager on a user's system. Failing to do so, it will fallback on a pure Ruby version that is guaranteed to work with any Ruby interpreter and on any platform.

## Contents

* [1. Usage](#1-usage)
* [2. API](#2-api)
  * [2.1 new](#21-new)
    * [2.1.1 :enabled](#211-enabled)
    * [2.1.2 :command](#212-command)
    * [2.1.3 :width](#213-width)
    * [2.1.4 :prompt](#214-prompt)
  * [2.2 page](#22-page)
  * [2.3 write](#23-write)
  * [2.4 try_write](#24-try_write)
  * [2.5 puts](#25-puts)
  * [2.6 close](#26-close)
  * [2.7 ENV](#27-env)

## 1. Usage

The **TTY::Pager** will pick the best paging mechanism available on your system when initialized:

```ruby
pager = TTY::Pager.new
```

Then to start paginating text call the `page` method with the content as the first argument:

```ruby
pager.page("Very long text...")
```

This will launch a pager in the background and wait until the user is done.

Alternatively, you can pass the `:path` keyword to specify a file path:

```ruby
pager.page(path: "/path/to/filename.txt")
```

If instead you'd like to paginate a long-running operation, you could use the block form of the pager:

```ruby
TTY::Pager.page do |pager|
  File.open("file_with_lots_of_lines.txt", "r").each_line do |line|
    # do some work with the line

    pager.write(line) # send it to the pager
  end
end
```

After block finishes, the pager is automatically closed.

For more control, you can translate the block form into separate `write` and `close` calls:

```ruby
begin
  pager = TTY::Pager.new

  File.open("file_with_lots_of_lines.txt", "r").each_line do |line|
    # do some work with the line

    pager.write(line) # send it to the pager
  end
rescue TTY::Pager::PagerClosed
  # the user closed the paginating tool
ensure
  pager.close
end
```

If you want to use a specific pager you can do so by invoking it directly:

```ruby
pager = TTY::Pager::BasicPager.new
# or
pager = TTY::Pager::SystemPager.new
# or
pager = TTY::Pager::NullPager.new
```

## 2. API

### 2.1 new

The `TTY::Pager` can be configured during initialization for terminal width, type of prompt when basic pager is invoked, and the pagination command to run.

For example, to disable a pager in CI you could do:

```ruby
pager = TTY::Pager.new(enabled: false)
````

#### 2.1.1 :enabled

If you want to disable the paging use the `:enabled` option set to `false`:

```ruby
pager = TTY::Pager.new(enabled: false)
```

This will directly print all the content to the standard output. If the output isn't a tty device, the pager will return the content directly to the caller.

#### 2.1.2 :command

To force `TTY::Pager` to always use a specific paging tool(s), use the `:command` option:

```ruby
TTY::Pager.new(command: "less -R")
```

The `:command` also accepts an array of pagers to use:

```ruby
pager = TTY::Pager.new(command: ["less -r", "more -r"])
```

If the provided pager command or commands don't exist on user's system, the pager will fallback automatically on a basic Ruby implementation.

To skip automatic detection of pager and always use a system pager do:

```ruby
TTY::Pager::SystemPager.new(command: "less -R")
```

#### 2.1.3 :width

Only the `BasicPager` allows you to wrap content at given terminal width:

```ruby
pager = TTY::Pager.new(width: 80)
```

This option doesn't affect the `SystemPager`.

To directly use `BasicPager` do:

```ruby
pager = TTY::Pager::BasicPager.new(width: 80)
```

#### 2.1.4 :prompt

To change the `BasicPager` page break prompt display, use the `:prompt` option:

```ruby
prompt = -> (page) { "Page -#{page_num}- Press enter to continue" }
pager = TTY::Pager.new(prompt: prompt)
```

### 2.2 page

To start paging use the `page` method. It can be invoked on an instance or a class.

The class-level `page` is a convenient shortcut. To page some text you only need to do:

```ruby
TTY::Pager.page("Some long text...")
````

You can also include extra initialization parameters. For example, if you prefer to use a specific command do this:

```ruby
TTY::Pager.page("Some long text...", command: "less -R")
````

The instance equivalent would be:

```ruby
pager = TTY::Pager.new(command: "less -R")
pager.page("Some long text...")
````

Apart from text, you can page file content by passing the `:path` option:

```ruby
TTY::Pager.page(path: "/path/to/filename.txt")
````

The final way is to use the class-level `page` with a block. After the block is done, the pager is automatically closed. For example, to read a file line by line with additional processing you could do:

```ruby
TTY::Pager.page do |pager|
  File.foreach("filename.txt") do |line|
    # do some work with the line

    pager.write(line) # write line to the pager
  end
end
```

The instance equivalent of the block version would be:

```ruby
pager = TTY::Pager.new
begin
  File.foreach("filename.txt") do |line|
    # do some work with the line

    pager.write(line) # write line to the pager
  end
rescue TTY::Pager::PagerClosed
ensure
  pager.close
end
```

### 2.3 write

To stream content to the pager use the `write` method.

```ruby
pager.write("Some text")
```

You can pass in any number of arguments:

```ruby
pager.write("one", "two", "three")
```

### 2.4 try_write

To check if a write has been successful use `try_write`:

```ruby
pager.try_write("Some text")
# => true
```

### 2.5 puts

To write a line of text and end it with a new line use `puts` call:

```ruby
pager.puts("Single line of content")
```

### 2.6 close

When you're done streaming content manually use `close` to finish paging.

All interactions with a pager can raise an exception for various reasons, so wrap your code using the following pattern:

```ruby
pager = TTY::Pager.new

begin
  # ... perform pager writes
rescue TTY::Pager::PagerClosed
  # the user closed the paginating tool
ensure
  pager.close
end
```

Alternatively use the class-level `page` call with a block to automatically close the pager:

```ruby
TTY::Pager.page do |pager|
  # ... perform pager writes
end
```

### 2.7 ENV

By default the `SystemPager` will check the `PAGER` environment variable. If the `PAGER` isn't set, the pager will try one of the searched commands like `less`, `more` or `pg`.

Therefore, if you wish to set your preferred pager you can either set up your shell like so:

```bash
PAGER=less -R
export PAGER
```

Or set `PAGER` in Ruby script:

```ruby
ENV["PAGER"]="less -R"
```

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/piotrmurach/tty-pager. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

1. Fork it ( https://github.com/piotrmurach/tty-pager/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

## Code of Conduct

Everyone interacting in the TTY::Pager project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/piotrmurach/tty-pager/blob/master/CODE_OF_CONDUCT.md).

## Copyright

Copyright (c) 2015 Piotr Murach. See LICENSE for further details.


CHANGELOG.md
# Change log

## [v0.14.0] - 2020-10-15

### Changed
* Change to update strings dependency
* Change to allow the pager to be specified via an absolute path (@WilliamMcCumstie)

### Fixed
* Fix name resolution with TTY::File

## [v0.13.0] - 2020-05-30

### Added
* Add support for streaming by Andrew Radev(@AndrewRadev)
* Add SystemPager integration tests by Andrew Radev(@AndrewRadev)

### Changed
* Change Pager structure and extract TTY::Pager::Abstract for the null, basic and
  system pagers
* Change BasicPager paging prompt to return only formatted prompt text and
  account for terminal height
* Change BasicPager to quit with a single keystroke
* Change BasicPager internals to increase readability
* Change SystemPager#command_exist? to search path
* Change tests structure for ease of maintenance by Andrew Radev(@AndrewRadev)
* Change gemspec to add metadata, remove test artefacts and dev dependencies
* Update strings & tty-screen dependencies
* Remove tty-which dependency

### Fixed
* Fix SystemPager to run command test silently via Open3 and thus work on JRuby

## [v0.12.1] - 2019-03-16

### Fixed
* Fix exception when use tty-pager with tty-file by @tiev

## [v0.12.0] - 2018-12-09

### Added
* Add SystemPager#run_command for running pager command silently and capturing output

### Changed
* Change SystemPager#new to allow specifying more than one custom command
* Change SystemPager#page to check if pager command can be run with flags before paging content
* Change to update tty-screen and tty-which dependencies

### Fixed
* Fix SystemPager to page content with ANSI codes

## [v0.11.0] - 2018-01-11

### Added
* Add `pg` and `most` to executables lookup

### Changed
* Change verse to strings dependency
* Change SystemPager to make all class methods public
* Change SystemPager#available to #find_executable
* Change SystemPager#available? to #exec_available
* Change Pager#find_available to #select_pager and make class method

### Fixed
* Fix Pager throwing exception if git is not installed by Katelyn Schiesser(@slowbro)

## [v0.10.0] - 2017-10-29

### Changed
* Change tty-screen dependency version
* Change gemspec to require Ruby >= 2.0.0
* Remove encoding comments

## [v0.9.0] - 2017-08-18

### Changed
* Change SystemPager to stop using fork, instead execute pager in subprocess
  and make it portable across platforms including Windows
* Change SystemPager to work on jruby
* Change NullPager to only print to stdout on tty device
* Change Pager to select SystemPager when paging command exists
* Remove jruby? checks from pager selection

## [v0.8.0] - 2017-07-14

### Added
* Add :command option to SystemPager to enforce choice of pagination tool
* Add Error type for specific error notifications

### Changed
* Change SystemPager to prevent initialization if pager isn't supported

### Fixed
* Fix BasicPager to take terminal width into account when displaying page break messages
* Fix SystemPager on Windows by detecting fork implementation

## [v0.7.1] - 2017-04-09

### Fixed
* Fix SystemPager raises error when executable is blank string by Jacob Evelyn (@JacobEvelyn)

## [v0.7.0] - 2017-03-20

### Changed
* Change files loading
* Update tty-which dependency

## [v0.6.0] - 2017-03-19

### Changed
* Change SystemPager to support piped git pagers by @JacobEvelyn

## [v0.5.0] - 2016-12-19

### Changed
* Change to call TTY::Which#exist? new api
* Change to send fork message directly to SystemPager
* Update tty-which
* Update verse dependency

## [v0.4.0] - 2016-02-06

### Changed
* Update tty-screen dependency

## [v0.3.0] - 2015-09-20

### Changed
* Change to use new tty-screen dependency

## [v0.2.0] - 2015-06-27

### Changed
* Change SystemPager to correctly paginate inside a process.

### [v0.1.0] - 2015-05-31

* Initial release

[v0.14.0]: https://github.com/piotrmurach/tty-pager/compare/v0.13.0...v0.14.0
[v0.13.0]: https://github.com/piotrmurach/tty-pager/compare/v0.12.1...v0.13.0
[v0.12.1]: https://github.com/piotrmurach/tty-pager/compare/v0.12.0...v0.12.1
[v0.12.0]: https://github.com/piotrmurach/tty-pager/compare/v0.11.0...v0.12.0
[v0.11.0]: https://github.com/piotrmurach/tty-pager/compare/v0.10.0...v0.11.0
[v0.10.0]: https://github.com/piotrmurach/tty-pager/compare/v0.9.0...v0.10.0
[v0.9.0]: https://github.com/piotrmurach/tty-pager/compare/v0.8.0...v0.9.0
[v0.8.0]: https://github.com/piotrmurach/tty-pager/compare/v0.7.1...v0.8.0
[v0.7.1]: https://github.com/piotrmurach/tty-pager/compare/v0.7.0...v0.7.1
[v0.7.0]: https://github.com/piotrmurach/tty-pager/compare/v0.6.0...v0.7.0
[v0.6.0]: https://github.com/piotrmurach/tty-pager/compare/v0.5.0...v0.6.0
[v0.5.0]: https://github.com/piotrmurach/tty-pager/compare/v0.4.0...v0.5.0
[v0.4.0]: https://github.com/piotrmurach/tty-pager/compare/v0.3.0...v0.4.0
[v0.3.0]: https://github.com/piotrmurach/tty-pager/compare/v0.2.0...v0.3.0
[v0.2.0]: https://github.com/piotrmurach/tty-pager/compare/v0.1.0...v0.2.0
[v0.1.0]: https://github.com/piotrmurach/tty-pager/compare/v0.1.0


lib/tty/pager/system.rb
# frozen_string_literal: true

require "open3"
require 'pathname'

require_relative "abstract"

module TTY
  module Pager
    # A system pager is used  on systems where native
    # pagination exists
    #
    # @api public
    class SystemPager < Abstract
      # Check if command exists
      #
      # @example
      #   command_exist?("less") # => true
      #
      # @param [String] command
      #   the command to check
      #
      # @return [Boolean]
      #
      # @api private
      def self.command_exist?(command)
        exts = ENV.fetch("PATHEXT", "").split(::File::PATH_SEPARATOR)
        if Pathname.new(command).absolute?
          ::File.exist?(command) ||
            exts.any? { |ext| ::File.exist?("#{command}#{ext}")}
        else
          ENV.fetch("PATH", "").split(::File::PATH_SEPARATOR).any? do |dir|
            file = ::File.join(dir, command)
            ::File.exist?(file) ||
              exts.any? { |ext| ::File.exist?("#{file}#{ext}") }
          end
        end
      end

      # Run pager command silently with no input and capture output
      #
      # @return [Boolean]
      #   true if command runs successfully, false otherwise
      #
      # @api private
      def self.run_command(*args)
        _, err, status = Open3.capture3(*args)
        err.empty? && status.success?
      rescue Errno::ENOENT
        false
      end

      # List possible executables for output paging
      #
      # The UNIX systems often come with "pg" and "more" but no "less" utility.
      # The Linux usually provides "less" and "more" pager, but often no "pg".
      # MacOS comes with "less" and "more" pager and no "pg".
      # Windows provides "more".
      # The "more" pager is the oldest utility and thus most compatible
      # with many systems.
      #
      # @return [Array[String]]
      #
      # @api private
      def self.executables
        [ENV["GIT_PAGER"], ENV["PAGER"], git_pager,
         "less -r", "more -r", "most", "pg", "cat", "pager"].compact
      end

      # Finds git pager configuration
      #
      # @api private
      def self.git_pager
        command_exist?("git") ? `git config --get-all core.pager` : nil
      end
      private_class_method :git_pager

      # Find first available termainal pager program executable
      #
      # @example Basic usage
      #   find_executable # => "less"
      #
      # @example Usage with commands
      #   find_executable("less", "cat")  # => "less"
      #
      # @param [Array[String]] commands
      #
      # @return [String, nil]
      #   the found executable or nil when not found
      #
      # @api public
      def self.find_executable(*commands)
        execs = commands.empty? ? executables : commands
        execs
          .compact.map(&:strip).reject(&:empty?).uniq
          .find { |cmd| command_exist?(cmd.split.first) }
      end

      # Check if command is available
      #
      # @example Basic usage
      #   available?  # => true
      #
      # @example Usage with command
      #   available?("less") # => true
      #
      # @return [Boolean]
      #
      # @api public
      def self.exec_available?(*commands)
        !find_executable(*commands).nil?
      end

      # Create a system pager
      #
      # @param [String] :command
      #   the command to use for paging
      #
      # @api public
      def initialize(command: nil, **options)
        super(**options)
        @pager_io = nil
        @pager_command = nil
        pager_command(*Array(command))

        if pager_command.nil?
          raise TTY::Pager::Error,
                "#{self.class.name} cannot be used on your system due to " \
                "lack of appropriate pager executable. Install `less` like " \
                "pager or try using `BasicPager` instead."
        end
      end

      # Send text to the pager process. Starts a new process if it hasn't been
      # started yet.
      #
      # @param [Array<String>] *args
      #   strings to send to the pager
      #
      # @raise [PagerClosed]
      #   strings to send to the pager
      #
      # @api public
      def write(*args)
        @pager_io ||= spawn_pager
        @pager_io.write(*args)
        self
      end
      alias << write

      # Send a line of text, ending in a newline, to the pager process. Starts
      # a new process if it hasn't been started yet.
      #
      # @raise [PagerClosed]
      #   if the pager was closed
      #
      # @return [SystemPager]
      #
      # @api public
      def puts(text)
        @pager_io ||= spawn_pager
        @pager_io.puts(text)
        self
      end

      # Stop the pager, wait for the process to finish. If no pager has been
      # started, returns true.
      #
      # @return [Boolean]
      #   the exit status of the child process
      #
      # @api public
      def close
        return true unless @pager_io

        success = @pager_io.close
        @pager_io = nil
        success
      end

      private

      # The pager command to run
      #
      # @return [String]
      #   the name of executable to run
      #
      # @api private
      def pager_command(*commands)
        if @pager_command && commands.empty?
          @pager_command
        else
          @pager_command = self.class.find_executable(*commands)
        end
      end

      # Spawn the pager process
      #
      # @return [PagerIO]
      #   A wrapper for the external pager
      #
      # @api private
      def spawn_pager
        # In case there's a previous pager running:
        close

        command = pager_command
        status = self.class.run_command(command)
        # Issue running command, e.g. unsupported flag, fallback to just command
        unless status
          command = pager_command.split.first
        end

        PagerIO.new(command)
      end

      # A wrapper for an external process.
      #
      # @api private
      class PagerIO
        def initialize(command)
          @command = command
          @io      = IO.popen(@command, "w")
          @pid     = @io.pid
        end

        def write(*args)
          io_call(:write, *args)
        end

        def puts(*args)
          io_call(:puts, *args)
        end

        def close
          return true if @io.closed?

          @io.close
          _, status = Process.waitpid2(@pid, Process::WNOHANG)
          status.success?
        rescue Errno::ECHILD, Errno::EPIPE
          # on jruby 9x waiting on pid raises ECHILD
          # on ruby 2.5/2.6, closing a closed pipe raises EPIPE
          true
        end

        private

        def io_call(method_name, *args)
          @io.public_send(method_name, *args)
        rescue Errno::EPIPE
          raise PagerClosed.new("The pager process (`#{@command}`) was closed")
        end
      end
    end # SystemPager
  end # Pager
end # TTY
